bsmetadata/preprocessing_tools/html_parser/filters_and_cleaners.py

```diff
@@ -0,0 +1,459 @@
+
+        return plain_text, self.metadata
+
+    def _br_conversion(self, tag):
+        if tag == "br":
+            self.text += "\n"
+
+    def _clean_relative_pos(self, metadata):
+        metadata_dict_idx = DefaultDict(dict)
+        for metadata_node in metadata:
+            metadata_dict_idx[metadata_node.char_start_idx][metadata_node.relative_start_pos] = (
+                "start",
+                metadata_node,
+            )
+            metadata_dict_idx[metadata_node.char_end_idx][metadata_node.relative_end_pos] = ("end", metadata_node)
+
+        for absolute_idx, value in metadata_dict_idx.items():
+            pos_sorted = sorted(list(value.keys()))
+            idx = 0
+            for pos in pos_sorted:
+                if metadata_dict_idx[absolute_idx][pos][0] == "start":
+                    metadata_dict_idx[absolute_idx][pos][1].relative_start_pos = idx
+                    idx += 1
+
+                if metadata_dict_idx[absolute_idx][pos][0] == "end":
+                    metadata_dict_idx[absolute_idx][pos][1].relative_end_pos = idx
+                    idx += 1
+
+    def _add_text(self, tag, new_text):
+        if tag in self.block_elements:
+            self.text = self._append_block_separator(self.text)
+        elif tag in INLINE_ELEMENTS_SPACING:
+            self.text = self._append_inline_element_separator(self.text)
+
+        if new_text:
+            self._append_text_content(new_text)
+
+        self._current_char_idx = len(self.text)
+
+    def _append_text_content(self, txt):
+        if self.current_tag == PRE_TAG:
+            self.text += txt
+        else:
+            txt = txt.replace("\u00a0", " ")
+
+            c = " "
+            if len(self.text) > 0:
+                c = self.text[-1]
+            for i in range(len(txt)):
+                c2 = txt[i]
+                if c2 == "\r" or c2 == "\n":
+                    c2 = " "
+                if not c.isspace() or not c2.isspace():
+                    self.text += c2
+                c = c2
+
+    def _append_block_separator(self, sb):
+        length = len(sb)
+        if length > 0:
+            # remove white space before paragraph break
+            # if self.last_tag != PRE_TAG:
+            #     while (length > 0 and sb[-1] == PLAIN_TEXT_SEPARATOR):
+            #         sb = sb[:-len(PLAIN_TEXT_SEPARATOR)]
+            if length > 0 and sb[-1] == PLAIN_TEXT_SEPARATOR:
+                sb = sb[:-1] + BLOCK_CONTENT_SEPARATOR
+            elif length > 0 and sb[-1] != BLOCK_CONTENT_SEPARATOR:
+                sb += BLOCK_CONTENT_SEPARATOR
+        return sb
+
+    def _append_inline_element_separator(self, sb):
+        length = len(sb)
+        if length > 0:
+            last_buffer_char = sb[-1]
+            if last_buffer_char != PLAIN_TEXT_SEPARATOR and last_buffer_char != BLOCK_CONTENT_SEPARATOR:
+                sb += PLAIN_TEXT_SEPARATOR
+        return sb
+
+    def _get_text_and_metadata(self, root):
```

Comment:
The name is a bit confusing given that this function only returns the text. Maybe `_get_text_and_update_metadata`?

Reply:
Change committed :slightly_smiling_face: